src/components/Table/hoc/withTableSettings/TableColumnSetup/DndTreeSelect.tsx

```diff
@@ -0,0 +1,139 @@
+import React from 'react';
+
+import {Grip} from '@gravity-ui/icons';
+import {DragDropContext, Draggable, Droppable} from 'react-beautiful-dnd';
+import type {OnDragEndResponder} from 'react-beautiful-dnd';
+
+import {useUniqId} from '../../../../../hooks';
+import {Icon} from '../../../../Icon';
+import {TreeSelect} from '../../../../TreeSelect/TreeSelect';
+import {TreeSelectItem} from '../../../../TreeSelect/TreeSelectItem';
+import type {TreeSelectItemProps} from '../../../../TreeSelect/TreeSelectItem';
+import type {RenderContainerType, RenderItem, TreeSelectProps} from '../../../../TreeSelect/types';
+import {ListContainerView} from '../../../../useList';
+import {reorderArray} from '../../../../useList/__stories__/utils/reorderArray';
+
+export type DndTreeSelectItem = TreeSelectItemProps & {
+    id: string;
+    isDragDisabled?: boolean;
+};
+
+export type RenderDndContainer = (container: {
+    renderList: () => React.JSX.Element;
+}) => React.ReactElement;
+
+const DEFAULT_RENDER_CONTAINER: RenderDndContainer = ({renderList}) => renderList();
+
+export type DndTreeSelectProps<T extends DndTreeSelectItem> = Omit<
+    TreeSelectProps<T>,
+    'items' | 'renderContainer' | 'renderItem'
+> & {
+    items: T[];
+    setItems: (_: T[]) => void;
+    renderContainer?: RenderDndContainer;
+    initialDroppableId?: string;
+};
+
+export function DndTreeSelect<T extends DndTreeSelectItem>({
+    items,
+    setItems,
+    initialDroppableId,
+    renderContainer: propsRenderContainer = DEFAULT_RENDER_CONTAINER,
+    ...treeSelectNativeProps
+}: DndTreeSelectProps<T>) {
+    const uniqId = useUniqId();
+
+    const renderContainer = React.useCallback<RenderContainerType<DndTreeSelectItem>>(
+        ({renderItem, visibleFlattenIds, items: _items, containerRef, id}) => {
+            const handleDrugEnd: OnDragEndResponder = ({destination, source}) => {
+                if (destination?.index !== undefined && destination?.index !== source.index) {
+                    const newItemsOrder = reorderArray(items, source.index, destination.index);
```

Comment:
Lets split logic: dnd and all the rest. This component will be responsible only for dnd and should know nothing about details of the list - it's items, styles, icons and so on. We will only pass items order and the way to change it (instead of items).

Reply:
I am not sure if i understand you correctly.

`This component will be responsible only for dnd and should know nothing about details of the list - it's items, styles, icons and so on` - seems there is no need in such component, cause `react-beautiful-dnd` is very simple and there is nothing you want to write once (I mean mega abstract dnd component for all kind of dnd tasks) to reuse in other places.

`DndTreeSelect` only for `withTableSettings` now (although I wrote it with the idea that a new component would appear in uikit). And cause it now only about `withTableSettings` and called `DndTreeSelect` it reuses `TreeSelect` and `TreeSelectItem` instead of custom properties.

Reply:
Ok, lets stick to this implementation, but let's not pass `items` directly, lets pass only items order.

Reply:
`let's not pass items directly, lets pass only items order.` - what do you mean ?